apps/safety/main.cpp

```diff
@@ -1,13 +1,219 @@
 #include "mbed.h"
+#include "CANMsg.h"
+#include "math.h"
+#include "rover_config.h"
 
+//configuration register 00h controls modes of chip
+// bit 15 - reset bit, wriitng 1 resets registers. bits 0-2 select operating mode
+// bus voltage register 02h
+// power register 03h
+//calibration register 05h
+// mask/enable register 06h
+//alert limit register 07h
+#define CONFIG_REGISTER 0x00
+#define VOLTAGE_REGISTER 0x01
+#define CURRENT_REGISTER 0x04
+
+// Datasheet for INA226 http://www.ti.com/lit/ds/symlink/ina226.pdf
+
+// leds on board
 DigitalOut led1(LED1);
+DigitalOut led2(LED2);
+DigitalOut led3(LED3);
+
+//Serial communications for debugging
+Serial pc(SERIAL_TX, SERIAL_RX, ROVER_DEFAULT_SERIAL_BAUD_RATE);
+
+//CAN initialization
+CAN can(CAN1_RX, CAN1_TX, ROVER_CANBUS_FREQUENCY); // dont know the actual CAN id's
+CANMsg rxMsg;
+//needs a specific CANID to update values
+const unsigned int expectedCANID = 0x000;
+
+//  I2C bus initialization
+I2C i2c(CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL); //pinnames found from pinnames.h
+const int SENSOR_SLAVE_ADDRESS = 0x1c << 1;
+
+enum configurationMode{
+    powerReset = 0,
+    triggered = 3,
+    continous = 7
+};
 
-const int k_interval_ms = 500;
+//below are settings for the INA226 chip. The default values have been selected 
+//but different conversion times can be selected as well. See datasheet for other settings
+const int busVoltageConversionTime = 4;
+const int shuntVoltageConversionTime = 4;
+const int averageMode = 0;
+
+// temperature conversion constants
+const float A = 0;
+const float B = 0;
+const float C = 0;
+const float rResistor = 0;
+
+//measurement thresholds, needs updating
+float currentUpperLim = 1;
+float currentLowerLim = 0;
+float voltageUpperLim = 1;
+float voltageLowerLim = 0;
+float tempUpperLim = 1;
+float tempLowerLim = 0;
+
+//funcs
+int inRange(float upperLim, float lowerLim, float val);
+float getCurrentData();
+float getVoltageData();
+float readThermosistor();
+int calibrateSensor();
+int configureSensor(configurationMode operationMode, int resetRegisters);
+void initCAN();
+void processCANMsg(CANMsg *p_newMsg);
 
 int main()
 {
+    // setup bus
+    i2c.frequency(10000);
+    initCAN();
+    // turn led off, turn on if issue occurs
+    led1 = 0;
+    led2 = 0;
+    led3 = 0;
+
     while (true) {
-        led1 = !led1;
-        wait_ms(k_interval_ms);
+        //read can messages to get data to configure device
+        u_int8_t command = 0;
+        if(can.read(rxMsg))
+        {
+            processCANMsg(&rxMsg);
```

Comment:
This is not good naming, by the name of "processCANMsg", it suggests that the function would already extract the data in the received message.